File: python/안전 영역(2468) copy.py
from collections import deque

n = int(input())
graph = [
    list(map(int, input().split()))
    for _ in range(n)
]

max_rain = 0
for i in range(n):
    max_rain = max(max_rain, max(graph[i]))

visited = [[0] * n for _ in range(n)]
dx = [-1,1,0,0]
dy = [0,0,-1,1]

cnt_arr = []

for i in range(1, max_rain):
    # 깊은 복사 
    rain_arr = list(map(list, graph))
    rain_visited = list(map(list, visited))
    for j in range(n):
        for k in range(n):
            # 최대 비 수보다 같거나 작으면 잠기는 지역
            if rain_arr[j][k] <= i:
                rain_arr[j][k] = 0
            else:
                rain_arr[j][k] = 1
    
    def bfs(x,y):
        queue = deque()
        queue.append((x,y))
        rain_visited[x][y] = 1

        while queue:
            x, y = queue.popleft()
            
            for a in range(4):
                nx = x + dx[a]
                ny = y + dy[a]

                if nx >=0 and nx < n and ny >=0 and ny < n and rain_visited[nx][ny] == 0 and rain_arr[nx][ny] == 1:
                    queue.append((nx,ny))
                    rain_visited[nx][ny] = 1
    
    cnt = 0
    for x in range(n):
        for y in range(n):
            if rain_arr[x][y] == 1 and rain_visited[x][y] == 0:
                bfs(x,y)
                cnt += 1
    cnt_arr.append(cnt)

print(max(cnt_arr) if cnt_arr else 1)

File: js/안전 영역(2468) copy.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const graph = input.slice(1).map((line) => line.split(' ').map(Number));

let maxRain = 0;
for (let i=0 ; i<n; i++) {
    maxRain = Math.max(maxRain, Math.max(...graph[i]));
}

const visited = Array.from({length:n}, () => Array(n).fill(false));

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

let cntArr = [];

for (let i=0; i<maxRain; i++) {
    // 원본 영역과 방문 배열 깊은 복사해서 사용
    const rainArr = graph.map(row => [...row]);
    const rainVisited = visited.map(row => [...row]);

    for (let j=0; j<n; j++) {
        for (let k=0; k<n; k++) {
            // 최대 빗물보다 영역높이가 작으면
            if (graph[j][k] <= i) {
                rainArr[j][k] = 0;
            } else {
                rainArr[j][k] = 1;
            }
        }
    }

    function bfs(x,y) {
        let queue = [];
        queue.push([x,y]);
        rainVisited[x][y] = true;
        let front = 0;

        while (front < queue.length) {
            const [x,y] = queue[front];
            front++;

            for (let a=0; a<4; a++) {
                let nx = x + dx[a];
                let ny = y + dy[a];

                if (nx >=0 && nx <n && ny >=0 && ny < n && !rainVisited[nx][ny] && rainArr[nx][ny] === 1) {
                    queue.push([nx,ny]);
                    rainVisited[nx][ny] = true;
                }
            }
        }
    }

    let cnt = 0;
    for (let w=0; w<n; w++) {
        for (let m=0; m<n; m++) {
            if (!rainVisited[w][m] && rainArr[w][m] === 1) {
                bfs(w,m);
                cnt++;
            } 
        }
    }
    cntArr.push(cnt);
}

const maxSafe = Math.max(...cntArr);
console.log(maxSafe === 0 ? 1 : maxSafe);